### PR TITLE
Refactor dataexchange controller, sub-controllers, tests

### PR DESCRIPTION
**Pull request by swashcap**
* **Problem:** Some of the Data Exchange code is messy. Some of it leaves us open to SQL injection attacks.
* **Solution:** _Refactor!_
* **Testing:** This doesn’t change functionality. All tests should pass.

### CODE REVIEW

File: packages/steelpenny/lib/controllers/dataexchange/request.js [hunk 1]
```diff
@@ -163,152 +152,178 @@ internals.makeTree = (items, parentItem) => {
 
 /*
  * Determines if the user should be able to edit a request
- *
  * @private
+ *
  * @param {Object} request object from database query
- * @param {Object} viewer's cookie object
```
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/request.js]: The [`@param` tag](http://usejsdoc.org/tags-param.html) doesn’t support type and description. Gotta put the parameter’s name in there.

File: packages/steelpenny/lib/controllers/dataexchange/request.js [hunk 2]
```diff
@@ -163,152 +152,178 @@ internals.makeTree = (items, parentItem) => {
...
+  {
+    json: { _id: requestId },
+  },
+  cookie
+) => Promise.all([
+  internals.getRequest(bookshelf, requestId),
+  internals.getGroups(bookshelf, requestId),
+  internals.getFilters(bookshelf, requestId),
+])
+  .then(([
```
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/request.js]: Another example of some extremely helpful destructuring. Don’t pluck those props when you can do this. Bonus: helps you visualize the shape of the argument(s).

File: packages/steelpenny/lib/controllers/dataexchange/request.js [hunk 3]
```diff
@@ -351,40 +365,47 @@ internals.mapSourceObjectValue = (value) => {
   return value;
 };
 
-const generateSourceObject = (bookshelf, query) => {
-  const parentId = query.json._topGroup._parentId;
-  const qryStr = 'SELECT * FROM dxdba.dx_generate_source_object_f(?);';
-
-  return bookshelf.knex.raw(qryStr, parentId)
```
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/request.js]: Knex.js docs always show bindings in an array. This probably works, but might as well follow their examples.

File: packages/steelpenny/test/integration/dataexchange.js [hunk 4]
```diff
@@ -5,33 +5,11 @@ const btoa = require('abab').btoa;
...
-function setCookie() {
-  return apiClient._keys.post({
-    username: btoa('mochatest'),
-    password: btoa('mochapassword'),
-  }).then((response) => {
-    const fullCookie = response.headers['set-cookie'][0];
-    cookie = fullCookie.slice(fullCookie.indexOf('=') + 1, fullCookie.indexOf(';'));
-  });
-}
+const filterStub = require('../stubs/dataexchange/filter.json');
```
Comment (swashcap) [inline packages/steelpenny/test/integration/dataexchange.js]: Static testing data doesn’t belong in _test/integration/_. Putting it in _test/stub/_ and separating by properties so it isn’t one giant, unreasonable blob.

File: packages/steelpenny/test/integration/dataexchange.js [hunk 5]
```diff
@@ -50,98 +28,105 @@ function getQueryJson() {
   };
 }
 
+const getDoRequest = (apiClient, cookie) => (objType, action, payload) =>
```
Comment (swashcap) [inline packages/steelpenny/test/integration/dataexchange.js]: Some more HOF sweetness. All tests involve _very similar_ `apiClient.dataexchange.get` calls: this DRYs tyings out a bit. See `doRequest` and its assignment below.

File: packages/steelpenny/test/integration/dataexchange.js [hunk 6]
```diff
@@ -50,98 +28,105 @@ function getQueryJson() {
...
+  apiClient.dataexchange.get(
+    objType,
+    action,
+    JSON.stringify(payload),
+    cookie
+  )
+  .then(({ data: { data: [result] } }) => result);
```
Comment (swashcap) [inline packages/steelpenny/test/integration/dataexchange.js]: Also, we’re only every interested in this mysteriously deeply nested result. Just return it!

File: packages/steelpenny/test/unit/dataexchange.js [hunk 7]
```diff
@@ -4,105 +4,125 @@ const chai = require('chai');
...
 const { internals, loadFromId } = require('../../lib/controllers/dataexchange/request');
+const { getIdExceptionMessage } = require('../../lib/controllers/dataexchange/common');
 const group = require('../../lib/controllers/dataexchange/group');
 const filter = require('../../lib/controllers/dataexchange/filter');
 
 // Set should property of all objects for BDD assertions
 chai.should();
 
-// helper function for mocking up bookshelf responses
-const bookShelfOverride = (returnObj) => { // eslint-disable-line arrow-body-style
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Don’t ignore linting rules unless there’s a good reason. This is easily written as:

```diff
const bookShelfOverride = (returnObj) => ({
  knex: {
    // ...
  },
});

File: packages/steelpenny/test/unit/dataexchange.js [hunk 8]
```diff
@@ -4,105 +4,125 @@ const chai = require('chai');
...
+
 describe('Dataexchange tests', () => {
   describe('Request controller', () => {
     describe('GET loadFromId', () => {
-      it('Returns error when request doesnt have any groups or filters', () => {
-        const bookshelf = null;
-        const query = { json: { _id: null } };
-        const cookie = null;
-
-        internals.getRequest = sinon.stub().returns({ rowCount: 1, rows: [{}] });
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: We’re stubbing internal methods by assignment without un-stubbing them at the end of our test. No good! If and when other tests rely upon these internal methods our stubs will still be in place. See [Sinon.js’s stub documentation](http://sinonjs.org/docs/#stubs) for more info, especially `sinon.Stub#restore`.

This is resolved by setting up stubs in the `before` hook and tearing them down in the `after` hook.

File: packages/steelpenny/test/unit/dataexchange.js [hunk 9]
```diff
@@ -4,105 +4,125 @@ const chai = require('chai');
...
+          cookie
+        )
         .catch((error) => {
           error.isBoom.should.equal(true);
           error.output.statusCode.should.equal(500);
           error.message.should.equal('result error');
-          error.data.should.equal('Delete function returned a different Group_id than what it was called with. Called: 1001 Returned: 999'); // eslint-disable-line max-len
-        });
-      });
+          error.data.should.equal(getIdExceptionMessage(999, 1001));
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Swapped for that _lib/controllers/dataexchange/common.js_ function‘s response. No more hard-coded strings!

File: packages/steelpenny/test/unit/dataexchange.js [hunk 10]
```diff
@@ -4,105 +4,125 @@ const chai = require('chai');
...
+        internals.getGroups.restore();
+        internals.getRequest.restore();
       });
     });
   });
 
   describe('Group controller', () => {
     describe('GET delete', () => {
-      it('Throws an error when the delete function returns a different id than expected', () => {
-        const bookshelf = bookShelfOverride({
```
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Variables only used once. Toss ‘em in.

File: packages/steelpenny/test/unit/dataexchange.js [hunk 11]
```diff
@@ -4,105 +4,121 @@ const chai = require('chai');
...
-        .catch((error) => {
-          error.isBoom.should.equal(true);
-          error.output.statusCode.should.equal(422);
-          error.output.payload.error.should.equal('Unprocessable Entity');
-          error.message.should.equal('invalid request details');
-          error.data.should.equal('returned groups: 0 filters: 0');
+        getFiltersStub.onCall(1).returns({
+          rowCount: 1,
+        });
+        getGroupsStub.onCall(1).returns({
```
Comment (drewlandis) [inline packages/steelpenny/test/unit/dataexchange.js]: Just by looking at this, it's hard to know that sinon's `.onCall(n)` method is zero-indexed (and their docs don't say it very explicitly either: [sinon.stubs](http://sinonjs.org/docs/#stubs)).  I would recommend changing lines 42 and 45 from `onCall(1)` --> `onSecondCall()` in order to make it more clear and more human-readable

File: packages/steelpenny/lib/controllers/dataexchange.js [hunk 12]
```diff
@@ -1,79 +1,108 @@
...
   filter,
+  group,
+  request,
 };
 
 module.exports.get = {
-  tags: ['api', 'sites'],
+  auth: false,
+  description: 'API for DataExchange front end',
+  notes: `Make \`GET\` requests to create, modify or delete a Data Exchange
```
Comment (drewlandis) [inline packages/steelpenny/lib/controllers/dataexchange.js]: nice!  Didn't realize you could do markdown-like code in swagger notes/

File: packages/steelpenny/lib/controllers/dataexchange.js [hunk 13]
```diff
@@ -1,79 +1,109 @@
...
         }
       } else {
-        parsedQuery = clientRequest.query;
+        requestData = {};
+      }
+
+      const method = subControllers[objType][action];
+
+      if (!method) {
+        throw boom.badRequest(`Action "${action}" not allowed`);
```
Comment (drewlandis) [inline packages/steelpenny/lib/controllers/dataexchange.js]: Smart 👍

File: packages/steelpenny/lib/controllers/dataexchange/filter.js [hunk 14]
```diff
@@ -1,161 +1,145 @@
...
+ *
+ * @param {Bookshelf} bookshelf Configured Bookshelf instance
+ * @param {Object} data Request's payload
+ * @returns {Promise<Object>}
+ */
+const getFormAttributes = (
+  bookshelf,
+  {
+    _modalityId,
+    _parent_id,
```
Comment (drewlandis) [inline packages/steelpenny/lib/controllers/dataexchange/filter.js]: is `_modalityId` meant to be camelCased and `_parent_id` meant to be underscore_cased?  Can we change them all to be camelCased?

File: packages/steelpenny/lib/controllers/dataexchange/filter.js [hunk 15]
```diff
@@ -1,161 +1,140 @@
...
-
-      // set filter ID
-      json._id = filterId;
-      json._attributes[0]._parentId = filterId;
-
-      // set filter statistics
-      json._statistics = filterStatistics;
-
-      return json;
+      const [_id, _statistics] = filter_id_statistics.split(':')
```
Comment (drewlandis) [inline packages/steelpenny/lib/controllers/dataexchange/filter.js]: yah... that is super clean.  Nice!

File: packages/steelpenny/lib/controllers/dataexchange/filter.js (hunk 14 above)
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/filter.js]: That’s data sent from the client. We’d have to adjust MICIS code.

File: packages/steelpenny/test/unit/dataexchange.js (hunk 11 above)
Comment (swashcap) [inline packages/steelpenny/test/unit/dataexchange.js]: Hm, didn’t know about `Stub#onSecondCall`. I’d it’s fine as-is, as folks familiar with Sinon.js should know what `onCall` does.

File: packages/steelpenny/lib/controllers/dataexchange.js (hunk 12 above)
Comment (rssk) [inline packages/steelpenny/lib/controllers/dataexchange.js]: same, super legit

File: packages/steelpenny/lib/controllers/dataexchange/common.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Get a function that executes a delete by ID.
+ *
+ * @param {string} functionName Database function's name
+ * @param {string} exceptionMessage Excerpt from database function's raised
+ * exception message
+ * @returns {Function}
+ */
+const getDelete = (functionName, exceptionMessage) => (
```
Comment (rssk) [inline packages/steelpenny/lib/controllers/dataexchange/common.js]: Interesting, I'm not toats sure if I understand how passed in args work, are they just put into the resturned funcs scope/closure?

File: packages/steelpenny/lib/controllers/dataexchange/filter.js [hunk 17]
```diff
@@ -1,161 +1,145 @@
...
-    throw error;
-  });
-};
+const deleteById = common.getDelete(
+  'dx_delete_data_request_template_filter_f',
+  'Cannot delete a filter belonging to a submitted request'
+);
 
 module.exports = {
+  delete: deleteById,
```
Comment (rssk) [inline packages/steelpenny/lib/controllers/dataexchange/filter.js]: neato

File: packages/steelpenny/lib/controllers/dataexchange/group.js [hunk 18]
```diff
@@ -70,55 +89,59 @@ const write = (bookshelf, query, cookie, passedId = null) => {
         _filters: filtersWithDetails,
       };
 
-      return json._topGroup ?
-        Object.assign({}, json, {
-          _topGroup: Object.assign({}, json._topGroup, group),
-        }) :
-        Object.assign({}, json, group);
+      return response._topGroup ?
+        merge({}, response, { _topGroup: group }) :
```
Comment (rssk) [inline packages/steelpenny/lib/controllers/dataexchange/group.js]: I'm not totally sure I understand this simplification

File: packages/steelpenny/lib/controllers/dataexchange/common.js (hunk 16 above)
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/common.js]: Yep, `getDelete` returns a `deleteById`-like function. This just captures the args for the returned function to use.

File: packages/steelpenny/lib/controllers/dataexchange/group.js [hunk 19]
```diff
@@ -135,19 +158,24 @@ const loadFilterStatistics = (bookshelf, query) => {
...
-          WHERE attribute_id = ${attributeId}
-            AND option_id = '${optionId}';
-        `;
         return Promise.all([
           filter,
           // USER_INPUT is for pre-approved data, user will get all data available
           // no need to get count
           optionLabel === 'USER_INPUT' ?
             undefined :
-            bookshelf.knex.raw(qryStr),
```
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/group.js]: Such hangovers.

File: packages/steelpenny/lib/controllers/dataexchange/group.js (hunk 18 above)
Comment (swashcap) [inline packages/steelpenny/lib/controllers/dataexchange/group.js]: `lodash#merge` is deep: it enumerates and assigns recursively through data structures.

File: packages/steelpenny/lib/controllers/dataexchange/common.js (hunk 16 above)
Comment (drewlandis) [inline packages/steelpenny/lib/controllers/dataexchange/common.js]: whoa... I had to look at this for a bit before I understood what it was doing.  Didn't realize you could do this.  Nice way to DRY things out 👍

File: packages/steelpenny/test/unit/dataexchange.js (hunk 8 above)
Comment (rssk) [inline packages/steelpenny/test/unit/dataexchange.js]: ☝️

**Pull request merged**